jigglypuffRL/deeprl/dqn/dqn.py

```diff
@@ -0,0 +1,239 @@
+    :param max_ep_len: (int) Maximum steps per episode
+    :param gamma: (float) discount factor
+    :param lr: (float) learing rate for the optimizer 
+    :param batch_size: (int) Update batch size
+    :param replay_size: (int) Replay memory size
+    :param plot_loss_reward_graph: (bool) To print the loss and reward as a graph or text
+    :param tensorboard_log: (str) the log location for tensorboard (if None, no logging)
+    :param seed : (int) seed for torch and gym
+    :param render : (boolean) if environment is to be rendered
+    :param device : (str) device to use for tensor operations; 'cpu' for cpu and 'cuda' for gpu
+    """
+
+    def __init__(
+        self,
+        network_type,
+        env,
+        double_dqn=False,
+        epochs=100,
+        max_iterations_per_epoch=100,
+        max_ep_len=1000,
+        gamma=0.99,
+        lr=0.001,
+        batch_size=32,
+        replay_size=100,
+        plot_loss_reward_graph=False,
+        tensorboard_log=None,
+        seed=None,
+        render=False,
+        device="cpu",
+    ):
+        self.env = env
+        self.double_dqn = double_dqn
+        self.max_epochs = epochs
+        self.max_iterations_per_epoch = max_iterations_per_epoch
+        self.max_ep_len = max_ep_len
+        self.replay_size = replay_size
+        self.lr = lr
+        self.gamma = gamma
+        self.batch_size = batch_size
+        self.plot_loss_reward_graph = plot_loss_reward_graph
+        self.tensorboard_log = tensorboard_log
+        self.render = render
+        self.loss_hist = []
+        self.reward_hist = []
+        self.max_epsilon = 1.0
+        self.min_epsilon = 0.01
+        self.epsilon_decay = 500
+
+        # Assign device
+        if "cuda" in device and torch.cuda.is_available():
+            self.device = torch.device(device)
+        else:
+            self.device = torch.device("cpu")
+
+        # Assign seed
+        if seed is not None:
+            torch.manual_seed(seed)
+            torch.backends.cudnn.deterministic = True
+            torch.backends.cudnn.benchmark = False
+            np.random.seed(seed)
+            self.env.seed(seed)
+            random.seed(seed)
+
+        # Setup tensorboard writer
+        self.writer = None
+        if self.tensorboard_log is not None:
+            from torch.utils.tensorboard import SummaryWriter
+
+            self.writer = SummaryWriter(log_dir=self.tensorboard_log)
+
+        self.create_model(network_type)
+
+    def create_model(self, network_type):
+        if network_type == "MLP":
+            self.model = MlpValue(
+                self.env.observation_space.shape[0],
+                self.env.action_space.n,
+                val_type="Qs",
+            )
+
+        if self.double_dqn and (network_type == "MLP"):
+            self.model = MlpValue(
+                self.env.observation_space.shape[0],
+                self.env.action_space.n,
+                val_type="Qs",
+            )
```

Comment:
To avoid code duplication and account for all cases, you could do something like 
```python
if network_type == 'MLP':
  self.q = MlpValue(...)
  self.q_t = deepcopy(Q)

  if self.double_dqn:
    self.q2 = MlpValue(...)
    self.q2_t = deepcopy(...)
```
This is still pretty tedious, but as of now it should be fine

Reply:
There will be an exponential in increase the number of models with more variants I suppose. I will account for all the cases using `self.model` and `self.target_model` probably.

Reply:
Don't worry, I've fixed in my next PR. Will be pushing now. Use that preferably

Reply:
I've added a get_model function which can be called directly from the algo class. It'll make things much more convenient hopefully

Reply:
See #57

Reply:
will take a look on this